File: scripts/preprocessing/gee/sentinel2_time_series.js
/*
 * ---
 * title: "Sentinel-2 Time Series Analysis"
 * author: "Brendan Casey"
 * created: "2024-12-18"
 * description: Generates a time series of Sentinel-2 satellite imagery,
 * calculates user-defined spectral indices, and outputs results as
 * multiband images for further analysis. 
 * ---
 */

/* 1. Setup
 * Prepare the environment, including the AOI, helper functions,
 * and date list for time series processing.
 */

/* Load helper functions */
var utils = require("users/bgcasey/science_centre:functions/utils");
var sentinelTimeSeries = require(
  "users/bgcasey/science_centre:functions/sentinel_time_series"
);
var sentinelIndicesAndMasks = require(
  "users/bgcasey/science_centre:functions/sentinel_indices_and_masks"
);

/* Define area of interest (AOI) */
var aoi = ee.FeatureCollection('FAO/GAUL_SIMPLIFIED_500m/2015/level1')
  .filter(ee.Filter.eq('ADM0_NAME', 'Canada'))
  .filter(ee.Filter.eq('ADM1_NAME', 'Alberta'))
  .geometry();

/* Small aoi for testing purposes */
// var aoi = ee.Geometry.Polygon([
//   [-113.5, 55.5],  // Top-left corner
//   [-113.5, 55.0],  // Bottom-left corner
//   [-112.8, 55.0],  // Bottom-right corner
//   [-112.8, 55.5]   // Top-right corner
// ]);

/* Create a date list 
 * The date list specifies the starting points for time 
 * intervals used to extract a time series. The createDateList 
 * function generates a list of dates at a specified interval 
 * (e.g., 1 year), beginning on the provided start date 
 * ('2000-06-01') and ending on the end date ('2024-06-01').
 *
 * For each date in the list, the s2_fn function will create a 
 * new end date by advancing the start date by a user-defined 
 * number of time units (e.g., 4 months, 6 weeks). Indices will 
 * be calculated for each of these time intervals.
 *
 * Due to memory limits when generating a time series of
 * Alberta wide images, time series are generated in two year
 * batches. Comment out the unused time periods.  
 */
 
var dateList = utils.createDateList(
  ee.Date('2019-06-01'), ee.Date('2020-06-01'), 1, 'years'
);

// var dateList = utils.createDateList(
//   ee.Date('2021-06-01'), ee.Date('2022-06-01'), 1, 'years'
// );

// var dateList = utils.createDateList(
//   ee.Date('2023-06-01'), ee.Date('2024-06-01'), 1, 'years'
// );

print("Start Dates", dateList);

/* Define reducer statistic */
var statistic = 'mean'; // Options: 'mean', 'median', 'max', etc.

/* 2. Sentinel-2 Time Series Processing
 * Calculate user-defined spectral indices for Sentinel-2 imagery.
 *
 * Available indices:
 * - CRE: Chlorophyll Red Edge Index
 * - DRS: Distance Red & SWIR
 * - DSWI: Disease Stress Water Index
 * - EVI: Enhanced Vegetation Index
 * - GNDVI: Green Normalized Difference Vegetation Index
 * - LAI: Leaf Area Index
 * - NBR: Normalized Burn Ratio
 * - NDRE1, NDRE2, NDRE3: Normalized Difference Red-Edge Indices
 * - NDVI: Normalized Difference Vegetation Index
 * - NDWI: Normalized Difference Water Index
 * - RDI: Ratio Drought Index
 */

// Generate Sentinel-2 Time Series
var s2 = sentinelTimeSeries.s2_fn(
  dateList, 121, 'days', aoi,
  ['CRE', 'DRS', 'DSWI', 'EVI', 'GNDVI', 'LAI', 'NBR', 
   'NDRE1', 'NDRE2', 'NDRE3', 'NDVI', 'NDWI', 'RDI']
)
  .map(function(image) { 
    return sentinelIndicesAndMasks.addNDRS(image, [210]); // Coniferous 
  })
  .map(function(image) { 
    return sentinelIndicesAndMasks.addNDRS(image, [220]); // Broadleaf 
  })
  .map(function(image) { 
    return sentinelIndicesAndMasks.addNDRS(image); // Mixedwood 
  })
  .map(function(image) {
    // Convert all bands in the image to float
    return image.toFloat();
  });

// print('Sentinel-2 Time Series:', s2);

/* 3. Check Calculated Bands
 * Review to make sure calculations and indices appear correct.
 */

// /* 3.1 Check band summary statistics
// * For each band calculate the min, max, and 
// * standard deviation of pixel values and print to console.
// * Check for values outside the expected range.
// */

// // Extract the first image in the time series
// var image_first = s2.first();

// // Calculate summary statistics for the first image
// var stats_first = image_first.reduceRegion({
//   reducer: ee.Reducer.min()
//     .combine(ee.Reducer.max(), '', true)
//     .combine(ee.Reducer.stdDev(), '', true),
//   geometry: aoi,
//   scale: 1000,
//   bestEffort: true,
//   maxPixels: 1e13
// });
// print('Summary Statistics for First Image:', stats_first);

// /* 3.2 Check Band Data Types
// * Bands need to be the same data type to export multiband rasters.
// */
// print("Band Names", image_first.bandNames()); 
// print("Band Types", image_first.bandTypes());

// /* 3.3 Plot NDVI
// * Visualize the NDVI for the first time step by adding it to the map.
// * Set visualization parameters to highlight vegetation health.
// */

// // Define visualization parameters for NDVI
// var ndviVisParams = {
//   min: -0.1, // Lower limit for NDVI values
//   max: 1.0,  // Upper limit for NDVI values
//   palette: ['blue', 'white', 'green'] // Color palette
// };

// // Extract the NDVI band from the first image
// var ndvi_first = image_first.select('NDVI');

// // Reduce resolution for visualization
// var ndvi_firstLowRes = ndvi_first.reproject({
//   crs: ndvi_first.projection(),
//   scale: 100
// });

// // Center the map on the area of interest (AOI)
// Map.centerObject(aoi, 9);

// // Add the low-resolution NDVI layer to the map
// Map.addLayer(ndvi_firstLowRes, ndviVisParams, 'NDVI (Low Res)');

/* 4. Export Sentinel-2 Time Series to Google Drive */
var folder = 'gee_exports';
var scale = 10; // Sentinel-2 resolution in meters
var crs = 'EPSG:4326'; // WGS 84 CRS

/* Define file naming function */
var fileNameFn = function(img) {
  var year = img.get('year').getInfo() || 'unknown';
  return 'sentinel2_multiband_' + year;
};

/* Export images to Google Drive */
utils.exportImageCollection(s2, aoi, folder, scale, crs, fileNameFn);

File: scripts/preprocessing/gee/python/sentinel2_time_series.py
# ---
# title:   Sentinel-2 Time Series Analysis
# author:  Brendan Casey
# created: 2026-07-10
# inputs:
#   - Sentinel-2 Surface Reflectance collection
#     (COPERNICUS/S2_SR_HARMONIZED)
#   - CA_FOREST_LC_VLCE2 land cover (for NDRS masks)
#   - FAO GAUL province boundaries (Alberta)
# outputs:
#   - Annual multiband spectral-index GeoTIFFs exported to
#     Google Drive at native (10 m) resolution
# notes:
#   Python port of sentinel2_time_series.js for the Earth
#   Engine Python API. Builds an annual date list, computes
#   user-selected spectral indices via the shared s2_fn
#   helper, adds NDRS bands for coniferous (210), broadleaf
#   (220), and mixedwood (all) forest, casts to Float32,
#   and exports multiband images.
#
#   The original Map.addLayer/Map.centerObject calls, vis
#   parameters, and debug print() blocks are omitted.
#
#   Deviation: the shared s2_fn helper expects a client-
#   side list of date strings, so the ee.List produced by
#   create_date_list is materialized with getInfo() before
#   being passed in.
#
#   Setup (once):
#     pip install earthengine-api
#     earthengine authenticate
#   Then set EE_PROJECT in _gee_config.py and run.
# ---

import os
import sys

import ee

# Make utils importable regardless of the working
# directory VS Code runs the script from
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from _gee_config import DRIVE_FOLDER
from utils import sentinel_indices_and_masks as indices
from utils.compute_report import ComputeReport
from utils.gee_helpers import (
    create_date_list,
    export_image_collection,
)
from utils.gee_utils import initialize_ee
from utils.sentinel_time_series import s2_fn

# 1. Setup ----

# 1.1 User parameters ----
S2_START_DATE = "2023-06-01"  # first time-series date
S2_END_DATE = "2024-06-01"  # last time-series date
S2_DATE_INTERVAL = 1  # step between series start dates
S2_DATE_INTERVAL_TYPE = "years"  # units for the step
S2_WINDOW = 121  # compositing window length
S2_WINDOW_TYPE = "days"  # units for the window
S2_INDICES = [
    "CRE", "DRS", "DSWI", "EVI", "GNDVI", "LAI", "NBR",
    "NDRE1", "NDRE2", "NDRE3", "NDVI", "NDWI", "RDI",
]
EXPORT_SCALE = 10  # native Sentinel-2 resolution (m)
EXPORT_CRS = "EPSG:4326"  # native export CRS
PRINT_STATS = True  # min/max check (slow for large AOIs)
USE_TEST_AOI = True  # True: small test AOI; False: Alberta
COMPUTE_REPORT = True  # write EECU usage report (txt)

# 1.2 Initialize Earth Engine ----
# Project ID is read from _gee_config.py
initialize_ee()

# 1.3 Set up compute usage report ----
# Profiles EECU usage per section. Best used with
# USE_TEST_AOI = True to find choke points cheaply.
report = ComputeReport(
    "sentinel2_time_series",
    out_dir=os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "gee_compute_reports",
    ),
    enabled=COMPUTE_REPORT,
)

# 2. Define study area ----
# Uses a small test polygon when USE_TEST_AOI is True;
# otherwise filters the FAO GAUL provinces for Alberta.

if USE_TEST_AOI:
    # Small aoi for testing purposes
    aoi = ee.Geometry.Polygon([
        [-113.5, 55.5],  # Top-left corner
        [-113.5, 55.0],  # Bottom-left corner
        [-112.8, 55.0],  # Bottom-right corner
        [-112.8, 55.5],  # Top-right corner
    ])
else:
    aoi = (
        ee.FeatureCollection(
            "FAO/GAUL_SIMPLIFIED_500m/2015/level1"
        )
        .filter(ee.Filter.eq("ADM0_NAME", "Canada"))
        .filter(ee.Filter.eq("ADM1_NAME", "Alberta"))
        .geometry()
    )

# 3. Build the time-series date list ----
# create_date_list returns an ee.List; s2_fn iterates a
# client-side list, so the dates are materialized as
# YYYY-MM-dd strings.
date_list = create_date_list(
    ee.Date(S2_START_DATE),
    ee.Date(S2_END_DATE),
    S2_DATE_INTERVAL,
    S2_DATE_INTERVAL_TYPE,
)
start_dates = (
    date_list.map(
        lambda d: ee.Date(d).format("YYYY-MM-dd")
    ).getInfo()
)

# 4. Sentinel-2 time-series processing ----
# Computes the selected spectral indices for each interval,
# adds NDRS bands for coniferous (210), broadleaf (220),
# and mixedwood (all forest) pixels, and casts to Float32.
s2 = s2_fn(
    start_dates,
    S2_WINDOW,
    S2_WINDOW_TYPE,
    aoi,
    S2_INDICES,
)
s2 = (
    s2.map(lambda img: indices.add_ndrs(img, [210]))
    .map(lambda img: indices.add_ndrs(img, [220]))
    .map(lambda img: indices.add_ndrs(img))
    .map(lambda img: img.toFloat())
)

# 5. Check calculated bands (optional) ----
# Earth Engine is lazy, so the profiler needs an evaluated
# computation to measure per-algorithm EECU usage.
if PRINT_STATS or COMPUTE_REPORT:
    reducer = (
        ee.Reducer.min()
        .combine(ee.Reducer.max(), "", True)
        .combine(ee.Reducer.stdDev(), "", True)
    )
    with report.section("Sentinel-2 first-image stats"):
        stats_first = (
            s2.first()
            .reduceRegion(
                reducer=reducer,
                geometry=aoi,
                scale=1000,
                bestEffort=True,
                maxPixels=1e13,
            )
            .getInfo()
        )
    print("Sentinel-2 first-image stats:", stats_first)

# 6. Export time series to Google Drive ----
# Exports each image in the collection as a multiband
# GeoTIFF, one export task per image.


def sentinel_file_name(img):
    """File name for the native-resolution export."""
    year = img.get("year").getInfo() or "unknown"
    return "sentinel2_multiband_" + str(year)


export_image_collection(
    s2,
    aoi,
    DRIVE_FOLDER,
    EXPORT_SCALE,
    EXPORT_CRS,
    sentinel_file_name,
)

# 7. Compute usage report ----
# Writes the profiled sections to gee_compute_reports/.
# Collection exports start many batch tasks, so per-task
# EECU totals are not logged here; monitor progress at
# https://code.earthengine.google.com/tasks
report.write()

# End of script ----
